File: test/spec/contracts/wuzzy-nest/searching.spec.ts
import { expect } from 'chai'

import {
  AOTestHandle,
  createLoader,
  OWNER_ADDRESS
} from '~/test/util/setup'

describe('Wuzzy-Nest Searching', () => {
  let handle: AOTestHandle

  beforeEach(async () => {
    handle = (await createLoader('wuzzy-nest')).handle
  })

  it('should validate search queries')

  it('should accept simple search queries', async () => {
    const indexType = 'ARNS'
    const arnsSubdomain = '@'
    const documents = [
      {
        arnsName: 'artbycity',
        contentType: 'text/plain',
        documentId: '67890',
        content: 'This is another test document from artbycity'
      },
      {
        arnsName: 'memeticblock',
        contentType: 'text/plain',
        documentId: '12345',
        content: 'This is a test document from memeticblock, it says test twice'
      },
      {
        arnsName: 'wuzzy',
        contentType: 'text/plain',
        documentId: '54321',
        content: 'This is yet another test document from wuzzy'
      }
    ]

    for (const doc of documents) {
      const result = await handle({
        From: OWNER_ADDRESS,
        Tags: [
          { name: 'Action', value: 'Index-Document' },
          { name: 'Index-Type', value: indexType },
          { name: 'Document-ARNS-Name', value: doc.arnsName },
          { name: 'Document-ARNS-Sub-Domain', value: arnsSubdomain },
          { name: 'Document-Content-Type', value: doc.contentType },
          { name: 'Document-Id', value: doc.documentId }
        ],
        Data: doc.content
      })
      if (result.Error) {
        console.log('Error indexing document for search test', result.Error)
      }
      expect(result.Error).to.not.exist
    }

    const searchOneResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Search' },
        { name: 'Query', value: 'memeticblock' }
      ]
    })
    if (searchOneResult.Error) {
      console.log(`DEBUG - AO Message Error: ${searchOneResult.Error}`)
    }
    expect(searchOneResult.Error).to.not.exist
    expect(searchOneResult.Messages).to.have.lengthOf(1)
    expect(searchOneResult.Messages[0].Data).to.exist
    const searchResults = JSON.parse(searchOneResult.Messages[0].Data)
    expect(searchResults.TotalCount).to.equal(1)
    expect(searchResults.Hits).to.have.lengthOf(1)
    expect(searchResults.Hits[0].doc.DocumentId).to.equal('12345')

    const searchTwoResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Search' },
        { name: 'Query', value: 'another' }
      ]
    })
    if (searchTwoResult.Error) {
      console.log(`DEBUG - AO Message Error: ${searchTwoResult.Error}`)
    }
    expect(searchTwoResult.Error).to.not.exist
    expect(searchTwoResult.Messages).to.have.lengthOf(1)
    expect(searchTwoResult.Messages[0].Data).to.exist
    const searchTwoResults = JSON.parse(searchTwoResult.Messages[0].Data)
    expect(searchTwoResults.TotalCount).to.equal(2)
    expect(searchTwoResults.Hits).to.have.lengthOf(2)
    const searchTwoHits = searchTwoResults.Hits.map(
      (hit: any) => hit.doc.DocumentId
    )
    expect(searchTwoHits).to.include('67890')
    expect(searchTwoHits).to.include('54321')

    const searchThreeResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Search' },
        { name: 'Query', value: 'nonexistent' }
      ]
    })
    if (searchThreeResult.Error) {
      console.log(`DEBUG - AO Message Error: ${searchThreeResult.Error}`)
    }
    expect(searchThreeResult.Error).to.not.exist
    expect(searchThreeResult.Messages).to.have.lengthOf(1)
    expect(searchThreeResult.Messages[0].Data).to.exist
    const searchThreeResults = JSON.parse(searchThreeResult.Messages[0].Data)
    expect(searchThreeResults.TotalCount).to.equal(0)
    expect(searchThreeResults.Hits).to.have.lengthOf(0)

    const searchFourResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Search' },
        { name: 'Query', value: 'test' }
      ]
    })
    if (searchFourResult.Error) {
      console.log(`DEBUG - AO Message Error: ${searchFourResult.Error}`)
    }
    expect(searchFourResult.Error).to.not.exist
    expect(searchFourResult.Messages).to.have.lengthOf(1)
    expect(searchFourResult.Messages[0].Data).to.exist
    const searchFourResults = JSON.parse(searchFourResult.Messages[0].Data)
    expect(searchFourResults.TotalCount).to.equal(3)
    expect(searchFourResults.Hits).to.have.lengthOf(3)
    // NB: memeticblock content says query twice, so should be ranked first
    const searchFourHits = searchFourResults.Hits.map(
      (hit: any) => hit.doc.DocumentId
    )
    expect(searchFourHits).to.include('12345')
    expect(searchFourHits).to.include('67890')
    expect(searchFourHits).to.include('54321')
  })

  it('should accept bm25 search queries', async () => {
    const indexType = 'ARNS'
    const arnsSubdomain = '@'
    const documents = [
      {
        arnsName: 'wuzzy1',
        contentType: 'text/plain',
        documentId: '11111',
        content: 'Wuzzy'
      },
      {
        arnsName: 'wuzzy2',
        contentType: 'text/plain',
        documentId: '22222',
        content: 'Wuzzy S'
      },
      {
        arnsName: 'wuzzy3',
        contentType: 'text/plain',
        documentId: '33333',
        content: 'Wuzzy AO Search'
      },
      {
        arnsName: 'wuzzy4',
        contentType: 'text/plain',
        documentId: '44444',
        content: 'Wuzzy Search'
      },
      {
        arnsName: 'wuzzy5',
        contentType: 'text/plain',
        documentId: '55555',
        content: 'Wuzzy Wuzzy Search Search'
      },
      {
        arnsName: 'wuzzy6',
        contentType: 'text/plain',
        documentId: '66666',
        content: 'Wuzzy Wuzzy Wuzzy Search Search Search'
      }
    ]

    for (const doc of documents) {
      const result = await handle({
        From: OWNER_ADDRESS,
        Tags: [
          { name: 'Action', value: 'Index-Document' },
          { name: 'Index-Type', value: indexType },
          { name: 'Document-ARNS-Name', value: doc.arnsName },
          { name: 'Document-ARNS-Sub-Domain', value: arnsSubdomain },
          { name: 'Document-Content-Type', value: doc.contentType },
          { name: 'Document-Id', value: doc.documentId }
        ],
        Data: doc.content
      })
      if (result.Error) {
        console.log('Error indexing document for search test', result.Error)
      }
      expect(result.Error).to.not.exist
    }

    const searchOneResult = await handle({
      From: OWNER_ADDRESS,
      Tags: [
        { name: 'Action', value: 'Search' },
        { name: 'Query', value: 'Wuzzy' },
        { name: 'Search-Type', value: 'bm25' }
      ]
    })
    if (searchOneResult.Error) {
      console.log(`DEBUG - AO Message Error: ${searchOneResult.Error}`)
    }
    expect(searchOneResult.Error).to.not.exist
    expect(searchOneResult.Messages).to.have.lengthOf(1)
    expect(searchOneResult.Messages[0].Data).to.exist
    const searchResults = JSON.parse(searchOneResult.Messages[0].Data)
    expect(searchResults.SearchType).to.equal('bm25')
    expect(searchResults.TotalCount).to.equal(6)
    expect(searchResults.Hits).to.have.lengthOf(6)
    expect(searchResults.Hits[0].doc.DocumentId).to.equal('11111')
  })
})

File: spec/wuzzy-nest/searching.spec.lua
local codepath = 'wuzzy-nest.wuzzy-nest'

describe('WuzzyNest Searching', function()
  _G.send = spy.new(function() end)
  local match = require('luassert.match')
  local utils = require('.utils')
  local WuzzyNest = require(codepath)
  before_each(function()
    CacheOriginalGlobals()
    WuzzyNest = require(codepath)
  end)
  after_each(function()
    RestoreOriginalGlobals()
    package.loaded[codepath] = nil
  end)

  local function addDocument(url, contentType, content)
    local handler = GetHandler('Index-Document')
    handler.handle({
      id = 'message-id-1',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Index-Document',
      data = content,
      ['document-last-crawled-at'] = tostring(os.time()),
      ['document-url'] = url,
      ['document-content-type'] = contentType
    })
  end

  it('requires query', function()
    _G.send = spy.new(function() end)
    local handler = GetHandler('Search')

    assert.has_error(function()
      handler.handle({
        id = 'message-id-1',
        from = _G.owner,
        target = 'wuzzy-nest-process-id',
        action = 'Search'
      })
    end)
  end)

  it('handles simple search queries', function()
    _G.send = spy.new(function() end)
    local handler = GetHandler('Search')
    local memeticblockUrl = 'https://memeticblock.com'
    addDocument(
      memeticblockUrl,
      'text/html',
      'This is a test document from memeticblock, it says test twice'
    )
    local artbycityUrl = 'https://artby.city'
    addDocument(
      artbycityUrl,
      'text/html',
      'This is another test document from artbycity'
    )
    local wuzzyUrl = 'https://wuzzy.io'
    addDocument(
      wuzzyUrl,
      'text/html',
      'This is yet another test document from wuzzy'
    )

    _G.send:clear()
    handler.handle({
      id = 'message-id-1',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Search',
      query = 'memeticblock'
    })
    local function is_simple_search_1(state, args)
      return function(msg)
        local results = require('json').decode(msg.data)
        return msg.target == _G.owner and
          msg.action == 'Search-Result' and
          results.SearchType == 'simple' and
          results.TotalCount == 1 and
          #results.Hits == 1 and
          results.Hits[1].doc.DocumentId == memeticblockUrl and
          results.Hits[1].score == 1
      end
    end
    assert:register('matcher', 'simple_search_1', is_simple_search_1)
    assert.spy(_G.send).was.called_with(match.is_simple_search_1())

    _G.send:clear()
    handler.handle({
      id = 'message-id-2',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Search',
      query = 'another'
    })
    local function is_simple_search_2(state, args)
      return function (msg)
        local results = require('json').decode(msg.data)
        local artbycityHit = utils.find(
          function (hit) return hit.doc.DocumentId == artbycityUrl end,
          results.Hits
        )
        local wuzzyHit = utils.find(
          function (hit) return hit.doc.DocumentId == wuzzyUrl end,
          results.Hits
        )
        return msg.target == _G.owner and
          msg.action == 'Search-Result' and
          results.SearchType == 'simple' and
          results.TotalCount == 2 and
          #results.Hits == 2 and
          results.Hits[1].score == 1 and
          results.Hits[1].doc.DocumentId == artbycityUrl and
          results.Hits[2].score == 1 and
          results.Hits[2].doc.DocumentId == wuzzyUrl
      end
    end
    assert:register('matcher', 'simple_search_2', is_simple_search_2)
    assert.spy(_G.send).was.called_with(match.is_simple_search_2())

    _G.send:clear()
    handler.handle({
      id = 'message-id-3',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Search',
      query = 'nonexistent'
    })
    local function is_simple_search_3(state, args)
      return function(msg)
        local results = require('json').decode(msg.data)
        return msg.target == _G.owner and
          msg.action == 'Search-Result' and
          results.SearchType == 'simple' and
          results.TotalCount == 0 and
          #results.Hits == 0
      end
    end
    assert:register('matcher', 'simple_search_3', is_simple_search_3)
    assert.spy(_G.send).was.called_with(match.is_simple_search_3())

    _G.send:clear()
    handler.handle({
      id = 'message-id-4',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Search',
      query = 'test'
    })
    local function is_simple_search_4(state, args)
      return function(msg)
        local results = require('json').decode(msg.data)
        return msg.target == _G.owner and
          msg.action == 'Search-Result' and
          results.SearchType == 'simple' and
          results.TotalCount == 3 and
          #results.Hits == 3 and
          results.Hits[1].score == 2 and
          results.Hits[1].doc.DocumentId == memeticblockUrl and
          results.Hits[2].score == 1 and
          results.Hits[2].doc.DocumentId == artbycityUrl and
          results.Hits[3].score == 1 and
          results.Hits[3].doc.DocumentId == wuzzyUrl
      end
    end
    assert:register('matcher', 'simple_search_4', is_simple_search_4)
    assert.spy(_G.send).was.called_with(match.is_simple_search_4())
  end)

  it('handles bm25 search queries', function()
    _G.send = spy.new(function() end)
    local handler = GetHandler('Search')
    local wuzzyUrl1 = 'https://wuzzy.io/doc1'
    addDocument(wuzzyUrl1, 'text/plain', 'Wuzzy')
    local wuzzyUrl2 = 'https://wuzzy.io/doc2'
    addDocument(wuzzyUrl2, 'text/plain', 'Wuzzy S')
    local wuzzyUrl3 = 'https://wuzzy.io/doc3'
    addDocument(wuzzyUrl3, 'text/plain', 'Wuzzy AO Search')
    local wuzzyUrl4 = 'https://wuzzy.io/doc4'
    addDocument(wuzzyUrl4, 'text/plain', 'Wuzzy Search')
    local wuzzyUrl5 = 'https://wuzzy.io/doc5'
    addDocument(wuzzyUrl5, 'text/plain', 'Wuzzy Wuzzy Search Search')
    local wuzzyUrl6 = 'https://wuzzy.io/doc6'
    addDocument(
      wuzzyUrl6,
      'text/plain',
      'Wuzzy Wuzzy Wuzzy Search Search Search'
    )

    _G.send:clear()
    handler.handle({
      id = 'message-id-1',
      from = _G.owner,
      target = 'wuzzy-nest-process-id',
      action = 'Search',
      query = 'Wuzzy',
      ['search-type'] = 'bm25'
    })
    local function is_bm25_search_1(state, args)
      return function(msg)
        local results = require('json').decode(msg.data)
        return msg.target == _G.owner and
          msg.action == 'Search-Result' and
          results.SearchType == 'bm25' and
          results.TotalCount == 6 and
          #results.Hits == 6 and
          results.Hits[1].doc.DocumentId == wuzzyUrl1 and
          results.Hits[1].score > 0.1 and  -- score = 0.10419692325373
          results.Hits[1].count == 1 and
          results.Hits[2].doc.DocumentId == wuzzyUrl2 and
          results.Hits[2].score > 0.09 and -- score = 0.097592892906662
          results.Hits[2].count == 1 and
          results.Hits[3].doc.DocumentId == wuzzyUrl6 and
          results.Hits[3].score > 0.08 and -- score = 0.092081001945156
          results.Hits[3].count == 3 and
          results.Hits[4].doc.DocumentId == wuzzyUrl5 and
          results.Hits[4].score > 0.08 and -- score = 0.089988251900948
          results.Hits[4].count == 2 and
          results.Hits[5].doc.DocumentId == wuzzyUrl4 and
          results.Hits[5].score > 0.07 and -- score = 0.084244320928547
          results.Hits[5].count == 1 and
          results.Hits[6].doc.DocumentId == wuzzyUrl3 and
          results.Hits[6].score > 0.07 and -- score = 0.077855004453629
          results.Hits[6].count == 1
      end
    end
    assert:register('matcher', 'bm25_search_1', is_bm25_search_1)
    assert.spy(_G.send).was.called_with(match.is_bm25_search_1())
  end)
end)
